### PR TITLE
Logging Activities to the Database

### PR DESCRIPTION
**Pull request by MinhLu12**
#28  -- THE LINK TO THE CONTEXT OF THE TICKET

The point of tracing is that we know when any user does an important activity like finishing a workflow or OAuthing and logging in. This would be really helpful for debugging purposes.

testtracing

The above picture is an example in our logs. This is tested.

### CODE REVIEW

File: database/dbService.js [hunk 1]
```diff
@@ -10,24 +10,34 @@ module.exports = {
...
+
+    insertActivityLog: function(ID, firstName, lastName, action) {
+        this.tryConnect().getConnection(function (err, con) {
+            var sql = queries.insertActivityLog;
+            con.query(sql, [ID, firstName, lastName, action], function (err, result) {
+                if (err) throw err;
+                con.release();
             });
         });
     },
```
Comment (MinhLu12) [inline database/dbService.js:28]: This is the function that'll insert the activity into the database.

We take in the user's ID, name and an action. Example of an action could be "User logged in successfully"

File: database/dbService.js [hunk 2]
```diff
@@ -10,24 +10,34 @@ module.exports = {
...
 
     replaceUserOnDuplicate: function(user, callback) {
         this.tryConnect().getConnection(function(err, con) {
-            var sql = queries.ReplaceUserOnDuplicate;
+            var sql = queries.replaceUserOnDuplicate;
             // Insert parameters
```
Comment (MinhLu12) [inline database/dbService.js:33]: Renamed for naming consistency

File: database/dbService.js [hunk 3]
```diff
@@ -10,24 +10,34 @@ module.exports = {
...
             con.query(sql, [user.id, user.googleID, user.gender, user.firstName, user.lastName, user.email, user.isProfileSetUp, user.location, user.phoneNumber,
                  // On Duplicate Key Update parameters
                  user.googleID, user.gender, user.firstName, user.lastName, user.email, user.isProfileSetUp, user.location, user.phoneNumber], 
                  function (err, result) {
-                con.release();
                 if (err) throw err;
-                return callback(result.insertId);
+                callback(result.insertId);
+                con.release();                
             });
```
Comment (MinhLu12) [inline database/dbService.js:41]: Realized we should release the connection back to the pool at the end of the query

File: database/queries/insert-activity-log.js [hunk 4]
```diff
@@ -0,0 +1,6 @@
+module.exports =
+
+`
+INSERT INTO activity_log (ID, FirstName, LastName, Action)
+VALUES (?, ?, ?, ?);
+`
```
Comment (MinhLu12) [inline database/queries/insert-activity-log.js]: Our string query for the insertion activity log

File: routes/userRoutes.js [hunk 5]
```diff
@@ -4,7 +4,10 @@ module.exports = app => {
     app.post('/api/user', (req, res) => {
 
         dbService.replaceUserOnDuplicate(req.body, function returnResponse(insertedId) {
-            if (insertedId === 0 || insertedId === req.body.id) {
+            var user = req.body; // Syntactic sugar
+            dbService.insertActivityLog(user.id, user.firstName, user.lastName, "User has finished the setup workflow");
+
+            if (insertedId === 0 || insertedId === user.id) {
                 return res.sendStatus(200);
```
Comment (MinhLu12) [inline routes/userRoutes.js]: This is tracing for when a user finishes the workflow

File: services/passport.js [hunk 6]
```diff
@@ -15,11 +15,15 @@ passport.use(
...
+                dbService.insertActivityLog(result[0].ID, result[0].FirstName, result[0].LastName, "Existing user finished OAuth and logged in");                
+
                 return done(null, result[0].ID);                
             }
             // If the user isn't OAuthed yet, insert the User model to our DB
             else {
                 dbService.insertUser(profile, function(id) {
+                    dbService.insertActivityLog(id, null, null, "New user OAuth'ed and logged in");                                    
                     done(null, id);
                 });
```
Comment (MinhLu12) [inline services/passport.js]: These two are OAuthing tracing for new and existing users.

File: services/passport.js [hunk 7]
```diff
@@ -15,11 +15,15 @@ passport.use(
...
+                // Trace it when the user logs in
+                dbService.insertActivityLog(result[0].ID, result[0].FirstName, result[0].LastName, "Existing user finished OAuth and logged in");                
+
                 return done(null, result[0].ID);                
             }
             // If the user isn't OAuthed yet, insert the User model to our DB
             else {
                 dbService.insertUser(profile, function(id) {
+                    dbService.insertActivityLog(id, null, null, "New user OAuth'ed and logged in");                                    
                     done(null, id);
```
Comment (MinhLu12) [inline services/passport.js]: Note here that Firstname and Lastname are null. We do not have access to these. When we insert the user, we only get back the user's ID.

I don't want to make another DB call to get the user via SQL and we could always find this user's name through the inserted ID.

Another alternate would be to use the "profile" variable for the first and last name but I'm not sure if that's 100% reliable or accurate. What if Google doesn't provide us their first or last names? there'd be another exception.

Comment (trevorhalvorson) [reply services/passport.js]: Do first and last names need to be associated with the logs? The user id alone is enough IMO if you just want to associate a user with an action

Comment (MinhLu12) [reply services/passport.js]: Ye, but it makes it more usable for people who look at the logs. Seeing a name would be helpful and very readable.

EX: "Oh, looks like Minh DID sign in but then logged out"

It eliminates the step of having to look up who logged in with another DB query (aka select * from user where ID = ?)

Comment (MinhLu12) [reply services/passport.js]: @trevorhalvorson If this is good enough lemme know

Comment (trevorhalvorson) [reply services/passport.js]: @MinhLu12 I still don't see the value in associating a full name with a log entry. What's the use case where that is needed?

**Review by trevorhalvorson — CHANGES_REQUESTED** [review]
first and last name info shouldn't be associated with logs

**Comment by MinhLu12** [comment]
@trevorhalvorson Made the changes and tested them. See the original post for the updated picture. I'm going to merge this now!

**Pull request merged**